komodo/matrix.py

```diff
@@ -24,10 +22,12 @@ def format_release(base, rhel_ver, py_ver):
 
 def get_matrix_base(release):
     """Return the base (e.g. matrix part of a concrete release). Should be the
-    inverse of format_release for actual, concrete matrix releases."""
-    for rhel_ver, py_ver in get_matrix():
-        suffix = format_release("", rhel_ver, py_ver)
-        if release.endswith(suffix):
-            return release.split(suffix, 1)[0]
+    inverse of format_release for actual, concrete matrix releases.
+    Hard-coded the suffix pattern '-py..-rhel.' or '-py...-rhel.'"""
```

Comment:
I wonder if we really need to hardcode details of our particular matrix into here? For example, can we define a more general pattern that ~all release names will adhere to?

Maybe releases will all look like `1111.11.11a11-pppp11-os11-other111`? At least that would avoid mentioning `py` and `rhel`. (Or maybe there's some other pattern, this is just an example.) The beauty of regex is that we can avoid having to be too specific. At its simplest, maybe it's just "everything before the first hyphen".

Alternatively, maybe the details of the matrix can go in a site config file? Or get passed in to the CLI that's making this query?

Reply:
Thanks for the comment. 
The format of a release is governed by [format_release](https://github.com/equinor/komodo/blob/f20735c60b84620ee3d08b15afd8f05331e30ddb/komodo/matrix.py#L19) function. I changed the code to keep the link with that function, which is a little better than pure hardcoded.
Nevertheless, I cannot see how we could get rid of hardcoding completely.

There is [one test](https://github.com/equinor/komodo/blob/f8fc14e46b0cfbc958d0ccb17a190dcbadd55fab/tests/test_matrix.py#L17) that would not pass with the `-` rule.

On the alternative you mention, do you suggest to have another argument to `transpile`? For example, we would have 4 arguments (or 2) for the current coordinates:  separate `os` from `os_version` and `prog_language` from `prog_language_version`?

Reply:
Eventually I think `format_release()` could look more like this:

```python
def format_release(base, matrix):
    for coords in product(*matrix.values()):
        coord_strings = [f"{name}{coord}" for name, coord in zip(matrix.keys(), coords)]
        yield '-'.join([base] + coord_strings)
```

Then be called like...

```python
matrix = {"py": ["3.8", "3.10", "3.11"], "rhel": ["7"], "loc": ["Azure", "Onprem"]}

for fname in format_release('baseybase', matrix):
    print(fname)
```

This produces:

```
baseybase-py3.8-rhel7-locAzure
baseybase-py3.8-rhel7-locOnprem
baseybase-py3.10-rhel7-locAzure
baseybase-py3.10-rhel7-locOnprem
baseybase-py3.11-rhel7-locAzure
baseybase-py3.11-rhel7-locOnprem
```

This way, it feels like we have lots more flexibility because the code never needs to know what our matrix names and coordinates are --- we could add Ruby or switch to Ubuntu without changing anything.

If this is all too much to ask of this PR, no worries --- I just wanted to clarify what I had in my mind! (For me as much as anything :)

Reply:
If this won't change, then I'll let you hit *Resolve conversation* to close this thread.

Reply:
That is indeed a good point. I renamed the entry argument as `matrix` and referred your specific suggestion in #332 . Will certainly help who picks that issue.